text/0024-schema-syntax.md

```diff
@@ -0,0 +1,449 @@
+
+## Basic example
+
+Here is the proposed syntax for the [TinyTodo schema](https://github.com/cedar-policy/cedar-examples/blob/main/tinytodo/tinytodo.cedarschema.json), whose auto-formatted JSON version is 160 lines long.
+```
+entity Application;
+entity User in [Team,Application] { name: String };
+entity Team in [Team,Application];
+entity List in [Application] {
+    owner: User,
+    name: String,
+    readers: Team,
+    editors: Team,
+    tasks: Set<{name: String, id: Long, state: String}>
+};
+
+action CreateList, GetLists
+    appliesTo { principal: [User], resource: [Application] };
+
+action GetList, UpdateList, DeleteList, CreateTask, UpdateTask, DeleteTask, EditShares
+    appliesTo { principal: [User], resource:[List] };
+```
+## Motivation
+
+Cedar schemas for non-toy applications are hard to read and write. The root cause of the difficulty is the use of JSON:
+
+* JSON has low information density. For three applications we’ve modeled — [TinyTodo](https://github.com/cedar-policy/cedar-examples/blob/main/tinytodo/tinytodo.cedarschema.json), [Document Cloud](https://github.com/cedar-policy/cedar-examples/blob/main/cedar-example-use-cases/document_cloud/schema.json), and [GitHub](https://github.com/cedar-policy/cedar-examples/blob/main/cedar-example-use-cases/github_example/schema.json) — schemas span 4-5 pages.
+* JSON does not support comments. This means that any design intent for a schema cannot be expressed inline.
+
+We believe that a custom syntax for Cedar schemas can help. It can be more information-dense, support comments, and leverage familiar syntactic constructs. We think customers will be happier with a custom syntax: we regularly hear favorable reviews of Cedar's custom policy syntax, as compared to a JSON-based syntax.
+
+## Detailed Design
+
+In what follows we first present the syntax by example. A full grammar is given at the end.
+
+### Comments
+
+Cedar schemas can use line-ending comments in the same style as Cedar, e.g., `// this is a comment`.
+
+### Namespace format
+
+```
+namespace My::Namespace {
+  ...
+}
+```
+
+A Schema can have zero or more namespace declarations. The contents of each (`...` above) comprise entity, action, and common-type declarations, all of which are unordered; i.e., they are all allowed to mutually refer to one another (but see discussion on common types below). The recommended style is to put entity declarations first, followed by action declarations, with common type declarations interspersed as needed. Entity, action, and common types referenced in the same namespace in which they are defined need not be prefixed by that namespace.
+
+Entity, action, and common-type declarations can appear on their own, outside of any namespace; in that case they are implicitly within the empty namespace.
+
+### Entity declarations
```

Comment:
The syntax and AST look very similar to Smithy and it would be a natural fit to model an API contract with the authorization domain model for it. Currently Smithy can model namespaces, generic data structures, and primitive types as well as `service`, `resource`, `operation` entities. We could define also something like `principals` and cedar traits to model the authz context type for the supported cedar operators on them.

An additional benefit of modelling the authz model for a cedar engine in smithy would be to also generate the SSDK stub for a service to fill in at runtime, exactly as the Smithy team does for the service payload.

I posit this would also help bringing up the SDF/CFN definitions and the reasoning efforts behind them.

So, have we considering such Smithy path? or is the intent to define a Cedar DSL and later bridge the two with a Smithy custom plugin to keep these two investments separate?

Reply:
I agree that this looks similar to Smithy.  I strongly recommend that consideration be given to consider using the Smithy IDL and perhaps contributing additional required traits/features/plugins.

Reply:
We met with the Smithy folks and, although there are ways for Cedar and Smithy to collaborate in the future, the Cedar schema isn't an IDL and there are enough differences between the goals of this syntax and the goals of the Smithy IDL that in my opinion we should proceed with this RFC as proposed.  Down the line we can discuss generating a Cedar schema from a Smithy API specification, but that would apply only for the particular case that Cedar is authorizing calls to a web API.

Reply:
Ack, thank you! Leaving out the reasoning aspects about a service posture that Smithy already models or what APIs Cedar is being used for, I'd argue that Cedar slices modelled with this doc's schema are an IDL, an authorization one if you will (and very powerful by the way). As a developer, I'd still have to create software types to marshall them when passing them over the wire to the service authorization subsystem and without supported typings I'll have to lift this undifferentiated work to bridge them with the entities of my application domain model already in Smithy, or Smithy-generated code for the language of choice, and which Cedar still requires to authorize on, and verify to get them right.

To be clear on the XY problem, given this schema syntax is to be its own DSL, are you considering improving dev-experince via code-generation tools for a polyglot service environment using Cedar?

Reply:
We are thinking about the dev-experience; Cedar already has its own [VSCode extension](https://github.com/cedar-policy/vscode-cedar/) which supports the current JSON format for schemas and will also support this new DSL.  The dev experience for using Cedar from any language other than Rust is currently subpar and we're well aware of that, but that is a more fundamental problem that extends beyond schema syntax.  Cedar simply doesn't currently have APIs in non-Rust languages that have functionality parity with its Rust API.

> As a developer, I'd still have to create software types to marshall them when passing them over the wire

This is true for some applications using Cedar; but some Cedar users never send Cedar data over the wire at all, they just call the Cedar library directly.

@mwhicks1 do you have comments on any of this?